release-21.1:  util/log: fix the long line splitting behavior with UTF-8 sequences

knz opened the pull request:
Backport 2/2 commits from #64900.

/cc @cockroachdb/release

---

Alleviates #61681.
Fixes #64896.

Release note (cli change): Long lines in the `crdb-v2` output
logging format are now split when they exceed 16K, instead
of 1K as previously.

Release note (bug fix): The automatic long line splitting behavior
of the new `crdb-v2` logging format now avoids splitting lines in the
middle of UTF-8 sequences, and avoids unbalanced redaction markers.

knz committed:
util/log: default the line split size to 16K instead of 1K

Release note (cli change): Long lines in the `crdb-v2` output
logging format are now split when they exceed 16K, instead
of 1K as previously.

knz committed:
util/log: fix line splitting behavior

Release note (bug fix): The automatic long line splitting behavior
of the new `crdb-v2` logging format now avoids splitting lines in the
middle of UTF-8 sequences, and avoids unbalanced redaction markers.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64914)
<!-- Reviewable:end -->

bdarnell reviewed (commented):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64914)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/util/log/format_crdb_v2.go, line 329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BkmzUD2Jb7LRhkKbA:-M_BkmzUD2Jb7LRhkKbB:b-46agif) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8fb121c661b6bd12ca42fb73a239ab5ff3256f8/pkg/util/log/format_crdb_v2.go#L329)):*
> ```Go
> // TODO(knz): This could also be made configurable by the user.
> //
> // NB: the value of this variable might be mentioned in the format's
> ```

"Might be"? Is it or not? 

If the limit is documented, it should also be documented that the effective limit is a little bit higher than this constant (by one fractional rune and an end redaction marker, if we want to expose the precise details)

Should the limit be specified in terms of bytes or runes? (either way, document which it is)
___
*[pkg/util/log/format_crdb_v2.go, line 387 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BhQfA3UsTwvmmY-Vg:-M_BhQfA3UsTwvmmY-Vh:bae682u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/format_crdb_v2.go#L387)):*
> ```Go
> 			// If we see an opening redaction marker, remember this fact
> 			// so that we close/open it properly.
> 			if strings.HasPrefix(msg[i:], startRedactionMarker) {
> ```

This looks relatively expensive, although on godbolt.org it looks like the compiler is optimizing away the copy implied by a string slice when it's passed straight to HasPrefix. We're leaning pretty heavily on the compiler's smarts here. It might be better to use a `range` loop over the string (which iterates over runes) instead of iterating over bytes and sometimes doing multi-byte stuff. (or for a larger refactoring, use strings.IndexAny instead of a byte-by-byte or rune-by-rune loop)
___
*[pkg/util/log/format_crdb_v2.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BlL2Q06nr2OnzI-As:-M_BlL2R5sS74rlz8MsE:b9i6x9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/format_crdb_v2.go#L403)):*
> ```Go
> 
> 		// Avoid breaking in the middle of UTF-8 sequences.
> 		_, width := utf8.DecodeRuneInString(msg[i:])
> ```

If we're calling DecodeRuneInString every byte, it's definitely going to be cheaper to just use a range loop over the string. But we could also avoid this check and only worry about multi-byte runes at the point where we decide to split the line. 
___
*[pkg/util/log/testdata/crdb_v2_break_lines, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_Bm9oRAUok_j88NJt3:-M_Bm9oRAUok_j88NJt4:b-a8w5ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/testdata/crdb_v2_break_lines#L61)):*
> ```
> ☃☃☃☃☃☃☃☃☃☃☃☃
> ----
> I000101 00:00:00.000000 0 :0 ⋮ [-]   ☃☃☃☃☃☃☃
> ```

I'm not sure what's going on, but these characters are rendering differently for me in reviewable (a small illegible blob) than the other snowmen. 


<!-- Sent from Reviewable.io -->

knz committed:
util/log: better honor the maximum line length

This way if/when we ever make it configurable and document it,
there won't be any surprises.

Release note: None

knz reviewed (commented):
I am replying to your comments in this reviewable session, but I have implemented the changes in the other PR - #64900 - let's move the convo over there.

I will update this backport PR after we're happy about the main PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-:-M_Kj-1h7mQpz1_Zpj1N:bgzeqy4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/util/log/format_crdb_v2.go, line 329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BkmzUD2Jb7LRhkKbA:-M_KZ8xrARALYMfDnmvt:b-cixm3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8fb121c661b6bd12ca42fb73a239ab5ff3256f8/pkg/util/log/format_crdb_v2.go#L329)):*
> "Might be"? Is it or not?

We have an agreement with the doc writers that they can edit the text above. So it may or may not be there any more. Hence the phrasing.

> Should the limit be specified in terms of bytes or runes? (either way, document which it is)

Bytes, good idea to document it. Done.

> it should also be documented that the effective limit is a little bit higher than this constant (by one fractional rune and an end redaction marker, if we want to expose the precise details)

I think we want this effective limit to be honored. Even though it currently *happens* not to be documented, we have requests to make it configurable (and thus become documented), and then it would likely turn into another security problem if folk truncate the data to fit in a buffer sized after the configured size.

Added a commit to enforce it (with tests).
___
*[pkg/util/log/format_crdb_v2.go, line 387 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BhQfA3UsTwvmmY-Vg:-M_KidIn4tZS53H43C5A:b-mgqgad) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/format_crdb_v2.go#L387)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This looks relatively expensive, although on godbolt.org it looks like the compiler is optimizing away the copy implied by a string slice when it's passed straight to HasPrefix. We're leaning pretty heavily on the compiler's smarts here. It might be better to use a `range` loop over the string (which iterates over runes) instead of iterating over bytes and sometimes doing multi-byte stuff. (or for a larger refactoring, use strings.IndexAny instead of a byte-by-byte or rune-by-rune loop)
</blockquote></details>

IMHO this is premature optimization - the common case is when there are no UTF-8 sequences, and that is covered by the conditional just above.
___
*[pkg/util/log/format_crdb_v2.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BlL2Q06nr2OnzI-As:-M_KimYbDy-s5ykG_gui:bh5a7tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/format_crdb_v2.go#L403)):*
That is what this condition above already is about:

> ```
> if markersStartWithSingleByteRune && msg[i] < utf8.RuneSelf { continue }
> ```
___
*[pkg/util/log/testdata/crdb_v2_break_lines, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_Bm9oRAUok_j88NJt3:-M_Kiv7mAoqL_YyPA_MV:bmlp26u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/testdata/crdb_v2_break_lines#L61)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm not sure what's going on, but these characters are rendering differently for me in reviewable (a small illegible blob) than the other snowmen. 
</blockquote></details>

![image.png](https://files.reviewableusercontent.io/images/reviewable/642886/I_6zh988pANUxwrL4QNC3rpF/image.png)

What do you have on your screen?

(I also checked in a hex dump that the utf-8 sequence is preserved.)


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64914)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/util/log/format_crdb_v2.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/64914#-M_BlL2Q06nr2OnzI-As:-M_KjlSD06FQUDv81OHu:b1410e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fa0b95da8669254ed5d9321a44488122f01674d/pkg/util/log/format_crdb_v2.go#L403)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That is what this condition above already is about:

> ```
> if markersStartWithSingleByteRune && msg[i] < utf8.RuneSelf { continue }
> ```

</blockquote></details>

(Although that condition was wrong, I just realized after replying to your comment. Fixed over in #64900.


<!-- Sent from Reviewable.io -->
